OpenRA.Mods.D2k/Traits/AttackSwallow.cs

```diff
@@ -75,21 +83,84 @@ public override Activity GetAttackActivity(Actor self, Target newTarget, bool al
 			return new SwallowTarget(self, newTarget, allowMove, forceAttack);
 		}
 
-		public class SwallowTarget : Attack
+		public class SwallowTarget : Activity
 		{
-			public SwallowTarget(Actor self, Target target, bool allowMovement, bool forceAttack)
-				: base(self, target, allowMovement, forceAttack) { }
+			readonly AttackSwallow attack;
+			readonly IMove move;
+			readonly IPositionable positionable;
+			readonly bool forceAttack;
+
+			protected Target target;
 
-			protected override Target RecalculateTarget(Actor self, out bool targetIsHiddenActor)
+			WDist minRange;
+			WDist maxRange;
+
+			public SwallowTarget(Actor self, Target target, bool allowMovement, bool forceAttack)
 			{
-				// Worms ignore visibility, so don't need to recalculate targets
-				targetIsHiddenActor = false;
-				return target;
+				this.target = target;
+				this.forceAttack = forceAttack;
+
+				attack = self.Trait<AttackSwallow>();
+				positionable = self.Trait<IPositionable>();
+
+				move = allowMovement ? self.TraitOrDefault<IMove>() : null;
 			}
 
-			protected override void DoAttack(Actor self, AttackFrontal attack, IEnumerable<Armament> armaments)
+			public override bool Tick(Actor self)
```

Comment:
Likewise here - most of the old `AttackFrontal` logic - ranges and facings - isn't actually needed for the sandworms.